### ISSUE TITLE
3.0 Release

### ISSUE THREAD

**Issue by timothyswt** [angular-ui/ui-grid#727]
Major architectural overhaul 

We have come up with a better architecture design for the grid and will be working here int he near future to bring you all the grid you and angularjs deserves.

More to come.

**Comment by prabhasoft** [comment]
Any release date pls as i have started developing a new project ?

**Comment by Cadrach** [comment]
Do you think a new RC is coming soon? Seems like some nice fixes are in already.

**Comment by PaulL1** [comment]
Our current aim is to fix all hi-pri bugs, then issue a new RC.  There are currently 17 hi-pri bugs outstanding.

**Comment by davidsanders** [comment]
Why is this? I thought ng-grid dev was ceasing as it turned into ui grid (http://ui-grid.info/).

Will these be the last fixes before development ends?

**Comment by PaulL1** [comment]
@davidsanders: This is on ui-grid / 3.0.  ng-grid development has effectively ceased, there hasn't been a new release for some time, and there are basically no contributors left on that version.

**Comment by davidsanders** [comment]
Ah, understood. I didn't realize ui grid was this project. I thought it had it's own project and this one was ending. A little odd, but understandable.

**Comment by uigridder** [comment]
Are we expecting a UI grid stable release anytime soon?

**Comment by PaulL1** [comment]
Really the driver is the hi-pri bugs.  If you look at that list, there are currently 8.  Grouping is now included, which was one of the last major functions we wanted for 3.0.  So we're reasonably close, but those remaining bugs are hard ones.

**Comment by deepu105** [comment]
Guys any idea if there will be a stable release before July?

**Comment by PaulL1** [comment]
Stable is probably relative.  What we have seems reasonably stable for the things that work, but there are known outstanding bugs.  We're ticking off the bugs, but help with that would make it more likely.  I'd probably have guessed yes that we'd get there, but to be honest things have been pretty slow in the last couple months.  We've actually fixed/modified a lot of things, all of which are improvements, but we're not making a lot of progress on the hi-pri, probably because they're either very difficult or obscure.

**Comment by deepu105** [comment]
any advice on using the current beta for production, in case release is not within july? anything to watch out for other than the open issues?

**Comment by PaulL1** [comment]
Just the open issues really.  I know people are using it, and I'm using it in my personal application (which isn't production but is at least alpha).  I think we may have some lurking browser compatibility problems, particularly with mobile and IE9, but those are in the issues.  

To some extent I'd just treat it like code that your team write - i.e. you test it and if it works for your test cases then you're good to go.  That means a bit more testing than you might normally do on a component, but in another sense any open source component probably gets a bit more testing than something you paid for and have support on.

Bottom line: if you've tested it for your scenarios and it works, and you don't blindly upgrade to new RCs, then you're probably fine.

**Comment by uigridder** [comment]
I am using ng-grid. We have a requirement in our project where we need to re position column A when column B is repositioned manually as these columns hold values that are logically related. Any ideas to get around doing this?

**Comment by PaulL1** [comment]
I think move columns triggers an event you can listen to.  You can then reorder the columnDefs, and call notifyDataChange. Away from my main computer or I could give more detail, but that's the basics.

**Comment by uigridder** [comment]
We have multiple lohical groupings of cols, so i believe it wud b our responsibility to reindex the cols... Right?

**Comment by PaulL1** [comment]
@uigridder, just checking - you mean ui-grid, not ng-grid, right?  Either way, yes, if you want to maintain logical groupings on the columns, but are letting users move individual columns around, then yes you'll have to detect that the move happened, and then move the remaining columns to maintain your logical grouping.

**Comment by uigridder** [comment]
Is there a way to programmatically reorder a column? ...something like columnToReorder.reorder(newColIndex). Because we have about 23-30 cols in our grid and 2-3 logically related groups each with 3-4 columns. I will go nuts trying to re-index them manually :((
Battle lost huh?..

**Comment by PaulL1** [comment]
Not easily.  Take a look at moveColumns - it has methods inside there that do move columns, and maybe it would be easier to submit a PR to expose those on the api than to write it yourself?

**Comment by uigridder** [comment]
We simulated a column move by using the code from the following events from the core file -onHeaderMouseDown and onHeaderDrop and providing the colToMoveIndex and targetIndex. These operations were done on ngGridEventColumns. However, the whole thing is executed each time we sort/filter/resize columns. I know this is a 'wontfix' now, but works for us.
Thanks for the help.

**Comment by PaulL1** [comment]
OK, I think we're getting close to 3.0 release.  I've created a new 3.1 milestone, and I'm working through all the issues to either leave them in 3.0 (implying we're not releasing until they're closed), put them in 3.1 (implying I think they're needed reasonably soon after 3.0), or putting them in future, implying that they're a good idea but really dependent on someone to decide to do a PR.

This should mean that you can tell how close 3.0 is just by watching the 3.0 milestone.  https://github.com/angular-ui/ng-grid/milestones

**Comment by deepu105** [comment]
:+1:  for moving enhancements to future

**Comment by rajesh0188** [comment]
When is the release? using ng-grid now in a project. Tried to update to rc-20 stable release some of the needed features for my project is not there.

**Comment by billyaraujo** [comment]
Was wondering the same thing. Would like to have the new features such as 'ui.grid.exporter' and  'ui-grid-grouping' that only exist in unstable. Also would like to update to angular 1.3.

**Comment by PaulL1** [comment]
https://github.com/angular-ui/ng-grid/milestones/3.0
94% complete apparently.

**Comment by billyaraujo** [comment]
What about a new RC so we can access some of the new features available in unstable?

**Comment by stoykostanchev** [comment]
RC 21 is out.

**Comment by rajesh0188** [comment]
Hello thanks for the update .... the product i am working on is about to release and we dont want to add unstable version to it ..... can you give me a release date so that my boss can decide if we wait for ui grid or find a different solution .....

**Comment by billyaraujo** [comment]
Thanks for the update. Really good to have these changes. Does this release still depend on angular 1.2? Because my bower file has:
  "dependencies": {
    "angular": "~1.2.16"
  },

**Comment by deepu105** [comment]
Will there be a release anytime soon guys?  Not that im complaining or anything, i know its an oss project, but still very annoying to see feature after feature addition without a stable release first :( would really love yo use this

**Comment by Thinkscape** [comment]
@deepu105 Well, that's exactly how it's supposed to go in OSS project :-)
You don't want new features and BC breaks **after** it's been released.

It's very close, though :+1: 
https://github.com/angular-ui/ng-grid/milestones/3.0

**Comment by deepu105** [comment]
@Thinkscape thanks for the update mate

**Comment by deepu105** [comment]
so there are 2 HP bugs and 2 smaller ones? :+1:  awesome

**Comment by rajesh0188** [comment]
Hi any news on release ....

**Comment by c0bra** [comment]
@rajesh0188 Still a couple bugs pending.

**Comment by billyaraujo** [comment]
Excellent work.

**Comment by sonthosh** [comment]
Is it released?

**Comment by billyaraujo** [comment]
Not yet. But with only 4 bugs pending shouldn't be long, I hope.

**Comment by deepu105** [comment]
@c0bra awesome guys, seems like all issues are fixed for 3.0? so can we expect a release soon?

**Comment by billyaraujo** [comment]
It has been released. Check the release tab.

**Comment by deepu105** [comment]
oops yes, :+1:  awesome

**Comment by wesleycho** [comment]
Closing, as this has been out for a little while.
